docs/DifferentiableProgramming.md

```diff
@@ -1477,10 +1487,112 @@ provided with a transpose (e.g. scalar multiplication, matrix transposition, and
 differential can be transposed.
 
-To make a function a linear map, use a `@transpose` attribute on the transpose
-function. If the function is only linear with respect to certain parameters, use
-a wrt: clause to indicate parameter positions, e.g. `@transpose(of: ..., wrt:
-(self, 0))`.
+To make an original function be linear, define a transpose function with a
+`@transpose` attribute that specifies the original function.
+
+##### Typing rules
+
+A function declaration does not have a fixed transpose type. This is because
+there can be multiple transpose functions that transpose the original function
+differently, e.g. with respect to different parameters, transposing under
+different generic constraints, etc.
+
+Given an original function declaration, a transpose function's type is
+determined from the following configurations:
+- Parameters to transpose with respect to.
+- Additional generic constraints that make the original function linear.
+
+The type of the transpose function under such configurations is a
+function that takes one argument whose type is the original function's result
+type and returns results that correspond to each original function parameter
+that is transposed with respect to. This definition, however, is a rough
+definition because there are differences among top-level functions, instance
+methods, and static methods.
+
+###### Linearity parameters
+
+Linearity parameters are parameters with respect to which a function is linear.
+The `@transpose` attribute accepts a `wrt:` argument which specifies a set of
+linearity parameters of the original function. If `wrt:` is not specified,
+linearity parameters default to all parameters. A `wrt:` argument in
+`@derivative` attributes can be a parameter index or a tuple of multiple
+parameter indices. All linearity parameters must have a type that conforms to
+both `Differentiable` and `AdditiveArithmetic` and satisfies
+`Self == Self.TangentVector`.
+
+When linearity parameters do not include all of the original function's
+parameters, those parameters must be taken in the front of the parameter list of
+the transpose function.
+
+The argument labels of original non-linearity parameters must be preserved in
+the transpose function. Other argument labels can be named freely. When there
+are multiple linearity parameters, it is useful to label the elements in the
+result tuple to distinguish between transposes with respect to different
+parameters.
+
+###### Top-level functions
+
+The type of the transpose function under such configurations is a function that
+takes the non-linearity parameters and a parameter whose type is the original
+function's result type and returns results that correspond to each original
+function parameter that is transposed with respect to. This definition, however,
+is a rough definition because there are differences among top-level functions,
+instance methods, and static methods.
```

Comment:
Seems like the same paragraph as on lines 1505-1510

Reply:
Deduped. Thanks!